server/src/main/java/org/elasticsearch/health/node/selection/HealthNodeSelectorTaskExecutor.java

```diff
@@ -0,0 +1,129 @@
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.elasticsearch.health.node.selection.HealthNodeSelector.TASK_NAME;
+
+/**
+ * Persistent task executor that is managing the {@link HealthNodeSelector}.
+ */
+public final class HealthNodeSelectorTaskExecutor extends PersistentTasksExecutor<HealthNodeSelectorTaskParams> {
+
+    private static final Logger logger = LogManager.getLogger(HealthNodeSelector.class);
+
+    private final ClusterService clusterService;
+    private final PersistentTasksService persistentTasksService;
+    private final AtomicReference<HealthNodeSelector> currentTask = new AtomicReference<>();
+
+    public HealthNodeSelectorTaskExecutor(Client client, ClusterService clusterService, ThreadPool threadPool) {
+        super(TASK_NAME, ThreadPool.Names.MANAGEMENT);
+        this.clusterService = clusterService;
+        persistentTasksService = new PersistentTasksService(clusterService, threadPool, client);
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, HealthNodeSelectorTaskParams params, PersistentTaskState state) {
+        HealthNodeSelector healthNodeSelector = (HealthNodeSelector) task;
+        currentTask.set(healthNodeSelector);
+        String nodeId = clusterService.localNode().getId();
+        logger.info("Node [" + nodeId + "] is selected as the current health node.");
+    }
+
+    @Override
+    protected HealthNodeSelector createTask(
+        long id,
+        String type,
+        String action,
+        TaskId parentTaskId,
+        PersistentTasksCustomMetadata.PersistentTask<HealthNodeSelectorTaskParams> taskInProgress,
+        Map<String, String> headers
+    ) {
+        return new HealthNodeSelector(id, type, action, getDescription(taskInProgress), parentTaskId, headers);
+    }
+
+    /**
+     * Returns the node id from the eligible health nodes
+     */
+    @Override
+    public PersistentTasksCustomMetadata.Assignment getAssignment(
+        HealthNodeSelectorTaskParams params,
+        Collection<DiscoveryNode> candidateNodes,
+        ClusterState clusterState
+    ) {
+        DiscoveryNode discoveryNode = selectLeastLoadedNode(clusterState, candidateNodes, node -> node.isMasterNode() == false);
+        if (discoveryNode == null) {
+            return NO_NODE_FOUND;
+        } else {
+            return new PersistentTasksCustomMetadata.Assignment(discoveryNode.getId(), "");
+        }
+    }
+
+    void startTask() {
+        persistentTasksService.sendStartRequest(
+            TASK_NAME,
+            TASK_NAME,
+            new HealthNodeSelectorTaskParams(),
+            ActionListener.wrap(r -> logger.debug("Created the health node selector task"), e -> {
+                Throwable t = e instanceof RemoteTransportException ? e.getCause() : e;
+                if (t instanceof ResourceAlreadyExistsException == false) {
+                    logger.error("failed to create health node selector task", e);
+                }
```

Comment:
Out of curiosity, why do we run into the `ResourceAlreadyExistsException`? Is this trying to re-allocate itself somewhere it already exists?

Reply:
Okay, that seems strange of the tasks framework to do to me. If we remove the check for that specific error, will this log on every node?

Reply:
yes, during start-up

Reply:
We could have the elected master node do this, but I think the same problem remains, we do not check if the task is there. We could potentially do that, right?

Reply:
If I understand right, it would be preferable to avoid doing anything that doesn't make sense to all nodes in a mixed cluster. It's ok if newer features don't work in a partially-upgraded cluster, see e.g. [these docs](https://www.elastic.co/guide/en/elastic-stack/8.2/upgrading-elasticsearch.html#rolling-upgrades):

> During a rolling upgrade, the cluster continues to operate normally. However, any new functionality is disabled or operates in a backward compatible mode until all nodes in the cluster are upgraded. New functionality becomes operational once the upgrade is complete and all nodes are running the new version. Once that has happened, there’s no way to return to operating in a backward compatible mode. Nodes running the previous version will not be allowed to join the fully-updated cluster.

The safest way to ensure that the cluster is fully upgraded is to check `clusterState.nodesIfRecovered().getMinNodeVersion()`. It's not enough to check the elected master, because this might not be the oldest node in the cluster.